feat(docs): setup RTD and add haproxy docs (#139)
canonical/ingress-configurator-operator: docs/how-to/haproxy-loadbalancing-grpc.md

---
myst:
  html_meta:
    "description lang=en": "How to use ingress-configurator to load balance a gRPC server through HAProxy."
---

(how_to_haproxy_loadbalancing_grpc)=

# How to load balance a gRPC server through HAProxy

This guide shows how to configure the `ingress-configurator` charm to expose a

> gRPC backend through HAProxy.

Reviewer: ```suggestion
You can configure the `ingress-configurator` charm to expose a
gRPC backend through HAProxy.
```
Nit to eliminate the meta-documentation in this sentence

gRPC load balancing requires the backend to support HTTP/2 over TLS. HAProxy
routes gRPC traffic through port 443 by default, or through a dedicated frontend
port configured via `external-grpc-port`.

## Prerequisites

- A HAProxy deployment with TLS configured. See the

>   [HAProxy getting started tutorial](https://charmhub.io/haproxy/docs) for setup

Reviewer: ```suggestion
  [HAProxy getting started tutorial](https://canonical.com/juju/docs/haproxy-charm/latest/tutorial/getting-started/) for setup
```
Another option is to set up an intersphinx mapping for the HAProxy docs and use the target `tutorial_getting_started` for the reference

  instructions.
- A gRPC backend reachable from the Juju model with TLS enabled. See the HAProxy
  operator's

>   [gRPC load balancing guide](https://canonical-haproxy-operator.readthedocs-hosted.com/en/latest/how-to/loadbalancing-for-a-grpc-server.html)

Reviewer: ```suggestion
  [gRPC load balancing guide](https://canonical.com/juju/docs/haproxy-charm/latest/how-to/loadbalancing-for-a-grpc-server/)
```
Same option here, can set up an intersphinx mapping for the HAProxy docs and use the target `how_to_loadbalancing_for_a_grpc_server`

  for instructions on setting up a gRPC backend (the `flagd` section).

## Deploy the `ingress-configurator` charm

```sh
juju deploy ingress-configurator grpc-configurator --channel=latest/edge
```

## Configure the charm

Set the backend address, port, protocol, and hostname. The `backend-protocol`
must be `https` for gRPC backends:

```sh
GRPC_SERVER_ADDRESS=<backend-ip>
juju config grpc-configurator \
    backend-addresses=$GRPC_SERVER_ADDRESS \
    backend-ports=<backend-port> \
    backend-protocol=https \
    hostname=<grpc-hostname>
```

If you want gRPC traffic on a dedicated frontend port rather than port 443, also
set `external-grpc-port`:

```sh
juju config grpc-configurator external-grpc-port=<port>

> ```

Reviewer: Nit: We could combine these steps into a single section, ``## Deploy and configure the `ingress-configurator` charm``

```sh
juju integrate grpc-configurator:haproxy-route haproxy

> ```

Reviewer: Nit: Is it absolutely necessary to use the `haproxy-route` endpoint for this goal? If yes, then I think we should point that out in this section. It's not just integrating the two charms, the magic happens by using that particular endpoint

juju integrate grpc-configurator:haproxy-route haproxy
```

## Verify the connection

Once all charms are active, verify the gRPC server is reachable through HAProxy:

```sh
HAPROXY_IP=$(juju status --format json | jq -r '.applications.haproxy.units."haproxy/0"."public-address"')
grpcurl -insecure -d '{}' <grpc-hostname>:<port> <service-method>

> ```

Reviewer: What does a successful verification look like?
